lib/analysis/support/messageGoal.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    guid,
+    isSlackMessage,
+    MessageOptions,
+    SlackFileMessage,
+} from "@atomist/automation-client";
+import {
+    actionableButton,
+    CommandHandlerRegistration,
+    doWithProject,
+    goal,
+    Goal,
+    PreferenceScope,
+    SdmContext,
+    slackInfoMessage,
+    slackSuccessMessage,
+} from "@atomist/sdm";
+import {
+    Action,
+    italic,
+} from "@atomist/slack-messages";
+import * as crypto from "crypto";
+import { PushMessage } from "../Interpretation";
+import { ProjectAnalyzer } from "../ProjectAnalyzer";
+
+/**
+ * Command to dismiss a certain PushMessage
+ */
+export const DismissMessageCommand: CommandHandlerRegistration<{ hash: string, msgId: string }> = {
+    name: "DismissMessage",
+    description: "Dismiss a Project Analysis message",
+    autoSubmit: true,
+    parameters: { hash: { displayable: false }, msgId: { displayable: false } },
+    listener: async ci => {
+        await ci.preferences.put(
+            `project-analysis.message.dismissed.${ci.parameters.hash}`,
+            true,
+            { scope: PreferenceScope.Sdm });
+        await ci.addressChannels(
+            slackSuccessMessage(
+                "Project Analysis",
+                "Successfully dismissed message"),
+            { id: ci.parameters.msgId });
+    },
+};
+
+/**
+ * Create a Goal that sends PushMessages.
+ * It adds a dismiss button to the messages it sends.
+ * @param analyzer
+ */
+export function messageGoal(analyzer: ProjectAnalyzer): Goal {
+    return goal({ displayName: "Messages" }, doWithProject(async gi => {
+        const { project } = gi;
+        const interpretation = await analyzer.interpret(project, gi, { full: false });
```

Comment:
Can we refactor this so that `messageGoal` doesn't know about the Analyzer so it's more generally usable. We could do this by passing in a second argument that is a function that knows how to find the messages.

The message type could be moved out of the Interpreter source file too.